source/_posts/2017/2017-09-04-how-apigen-survived-its-death.md

```diff
@@ -0,0 +1,202 @@
+    <br><br>
+    I got frustrated. I spent year on a project that is still not working out of the box. So I took spring off to change it. <strong>My goal was to replace reflection or let the project die in peace</strong>.
+    <br><br>
+    This is story about the whole journey of ups and downs.
+'''
+tweet: "How ApiGen Survived its Death #php #apigen #phpdoc #teamwork"
+---
+
+Prepare for deep darkness, (almost) burning out and... team work that helped me to make it to the end.
+
+<img src="/assets/images/posts/2017/apigen-revival/apigen.png" class="img-thumbnail">
+
+
+## Step 1: Bump to PHP 7.1
+
+I love PHP 7.1 and I use it everywhere since it's release in 2016. This year [more and more big projects are migrating]([https://www.tomasvotruba.cz/blog/2017/06/05/go-php-71/), yet this is still low % of all packages.
+
+So my condition was to [bump minimal requirement to PHP 7.1](https://github.com/ApiGen/ApiGen/issues/779#issuecomment-285960383). Current maintainers agreed, so I had green on. Thank you [Alexander Jank](https://github.com/jankal) for your support in my first PRs to ApiGen this year. 
+
+
+## Step 2: Pick the right Reflection Successor
+
+ApiGen uses reflection to analyze classes, their methods, interfaces, traits, parent class of the class, their methods etc.
+
+**Pure PHP reflection can be barely used for advanced tasks like the last one**, so I'd have to rewrite whole package myself. That's not a way to go, if you want to have a calm life and normal sleep.
+
+The question was, **where to find the right one?**
+
+I knew few, but not any that would be able to parse PHP 7.1 by itself. I was aware of [nikic/php-parser](https://github.com/nikic/PHP-Parser), that was maintained and future-compatible (Nikic even added PHP 7.2 features recently). But it was only parsing tool, not smart reflection wrapper.
+
+
+### Reaching out for Help
+
+In that time, I came across (Projects using the PHP Parser)[https://github.com/nikic/PHP-Parser/wiki/Projects-using-the-PHP-Parser] in Wiki of PHP-Parse, I consulted with [Jan Tvrdik](https://github.com/jantvrdik) and [https://twitter.com/omirtes](Ondrej Mirtes). 
+
+This all [lead me to a package](https://github.com/ApiGen/ApiGen/issues/817) called [Roave/BetterReflection](https://github.com/roave/better-reflection) by [James Titcumb](https://www.jamestitcumb.com/) and [Marco Pivetta](https://ocramius.github.io/).
+ 
+I'm bit suspicious to projects that were lastly tagged a half year ago, but I felt I could gave it a go.
+
+
+## Step 3: Replace Reflection
+
+All right, package picked! The <strike>fun</strike> hell was about to begin.
+
+Imagine your whole application uses everywhere a package, that got stuck 4 PHP versions ago.You feel it more and more. Everyday you need to patch it, because there are other packages that are up-to-date. You know, like PHP itself.
```

Comment:
missing space after `ago.`

Reply:
I would remove the comma after "patch it".